avm/src/vm/memory.rs

```diff
@@ -1,654 +1,390 @@
 use std::convert::TryInto;
-use std::fmt;
-use std::slice;
 use std::str;
 
-use regex::Regex;
-
 use crate::vm::program::Program;
 
 // -2^63
 pub const CLOSURE_ARG_MEM_START: i64 = -9223372036854775808;
-pub const CLOSURE_ARG_MEM_SIZE: usize = 4;
+pub const CLOSURE_ARG_MEM_END: i64 = CLOSURE_ARG_MEM_START + 3;
 // Flags for the registers_ish vector. The normal address flag indicates that the data is stored
 // normally in either the memory or fractal memory structures. The fixed pointer address flag
 // indicates that the value in the memory structure is actually a pointer to an i64 value. The
 // handlermemory pointer address flag indicates that the value in the memory structure is actually
 // a pointer to a HandlerMemory object.
 const NORMAL_ADDR: i8 = 0;
-const FX_PTR_ADDR: i8 = 1;
-const HM_PTR_ADDR: i8 = 2;
+const GMEM_ADDR: i8 = 1;
+const ARGS_ADDR: i8 = 2;
+
+fn addr_type(addr: i64) -> i8 {
+  return if addr >= 0 {
+    NORMAL_ADDR
+  } else if addr <= CLOSURE_ARG_MEM_END {
+    ARGS_ADDR
+  } else {
+    GMEM_ADDR
+  }
+}
 
 /// Memory representation of a handler call
 #[derive(Clone)]
+#[derive(Debug)]
 pub struct HandlerMemory {
-  /// Global memory reference
-  gmem: &'static Vec<u8>,
-  /// Memory of the handler for fixed size data types
-  pub mem: Vec<i64>,
-  /// Fractal memory storage for variable-length data types to an instance of the same struct
-  pub fractal_mem: Vec<HandlerMemory>,
-  /// Helper for fractal_mem to lookup the actual location of the relevant data, since instantiating
-  /// HandlerMemory instances for each memory usage need is expensive. -1 means the data is in mem,
-  /// otherwise it's mapping to the index in fractal_mem that houses the relevant data.
-  pub either_mem: Vec<i64>,
-  /// Flag indicating the memory stored is actually a pointer
-  registers_ish: Vec<i8>,
-  /// Temporary hack
-  pub is_fixed: bool,
-  /// Memory space used for closure arguments
-  closure_args: Vec<HandlerMemory>, // Option doesn't work for some reason
+  /// The set of memory blocks. Each block consists of a tuple of two values. If the first value is
+  /// zero then the second value is the actual data, if it is non-zero then the two values together
+  /// represent a virtual pointer to another memory block and value (if the second value is zero,
+  /// it represents a pointer to a nested array of data, otherwise a pointer to an explicit value)
+  mems: Vec<Vec<(usize, i64)>>,
+  /// The address spaces for the handler memory. The first  is the "normal" memory space, and the
+  /// second is the args memory space.
```

Comment:
So, you don't have to explicitly declare a `HandlerMemory` is for a closure anymore, and I also use this space as a "scratch" space for certain opcodes, so that's why I don't mention "closure" in the comments.